### ISSUE TITLE
Guidelines For Rvalue References In APIs

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8714]
Guidelines For Rvalue References In APIs<br>
<br>
https://ift.tt/Kg7evMn<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<article>
<header>
<h1>Guidelines For Rvalue References In APIs</h1>
</header>
<p>I&rsquo;ll be giving a talk at <a href="https://conference.accu.org/2018/sessions.html#XIjustwantedtopointtosomething">ACCU</a> about when to use which pointer types and why.</p>
<p>While working on that I made some guidelines for rvalue references in interfaces which didn&rsquo;t quite fit the talk, so I&rsquo;m writing about them here.</p>
<p>When should you use rvalue references as function parameters?</p>
<p>When as return types?</p>
<p>What are ref-qualified member functions and when and how should you use them?</p>
<p>Let&rsquo;s tackle it one by one.</p>
<h2>Rvalue Reference Parameters</h2>
<p>Some functions take an rvalue reference:</p>
<p>Note that this function does not take an rvalue reference:</p>
<div>
<pre tabindex="0"><code><span><span><span>template</span> <span>&lt;</span><span>typename</span> <span>T</span><span>&gt;</span>
</span></span><span><span><span>void</span> <span>foo</span><span>(</span><span>T</span><span>&amp;&amp;</span> <span>param</span><span>);</span>
</span></span></code></pre>
</div>
<p>Because <code>T</code> is a template parameter of the function, different rules kick in transforming it into a <em>forwarding reference</em>. This is a hack used to fulfill a completely different purpose, perfect forwarding. Read more about it <a href="https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers">here</a>, for example.</p>
<blockquote>
<p>Note: They&rsquo;re still called &ldquo;universal references&rdquo; in that post; they have been renamed since.</p>
</blockquote>
<p>So why do functions take an rvalue reference?</p>
<p>I&rsquo;ve identified three use cases: one common, one rare and one experimental that may be useful.</p>
<h3>1. Rvalue + Lvalue Reference Parameter Combination For Performance</h3>
<p>The best example of this is <code>std::vector&lt;T&gt;::push_back()</code>:</p>
<div>
<pre tabindex="0"><code><span><span><span>void</span> <span>push_back</span><span>(</span><span>const</span> <span>T</span><span>&amp;</span> <span>obj</span><span>);</span>
</span></span><span><span><span>void</span> <span>push_back</span><span>(</span><span>T</span><span>&amp;&amp;</span> <span>obj</span><span>);</span>
</span></span></code></pre>
</div>
<blockquote>
<p>Note that even though <code>T</code> is a template parameter it is a template parameter of the <em>class</em>, not the <code>push_back()</code> function. This means that <code>T&amp;&amp;</code> is still an rvalue reference.</p>
</blockquote>
<p>When you have a parameter where you want to store in a different place, a good guideline is to pass that parameter by value and move it into the final location.</p>
<p>The classical example of it is a constructor:</p>
<div>
<pre tabindex="0"><code><span><span><span>explicit</span> <span>person</span><span>(</span><span>std</span><span>::</span><span>string</span> <span>name</span><span>)</span>
</span></span><span><span><span>:</span> <span>name_</span><span>(</span><span>std</span><span>::</span><span>move</span><span>(</span><span>name</span><span>))</span>
</span></span><span><span><span>{}</span>
</span></span></code></pre>
</div>
<p>That way when you pass an lvalue, you pay for a copy (into the parameter) and a move (into the final location), and when you pass an rvalue, you pay for a move (into the parameter) and a move (into the final location).</p>
<p>Taking a <code>const T&amp;</code> would have a copy (into the final location) for both lvalues and rvalues, and passing <code>T&amp;&amp;</code> only would prevent passing lvalues. So with <code>T</code> both works and because moves are <a href="https://www.foonathan.net/blog/2017/09/14/destructive-move.html">generally</a> cheap, there isn&rsquo;t much downside.</p>
<p>However the perfect solution would copy lvalues and move rvalues, so you don&rsquo;t have to pay for the additional move from the parameter into the final location. In order to achieve that, you have to overload your function for both lvalues and rvalues.</p>
<p>But this leads to <code>2^n</code> duplicated functions, where <code>n</code> is the number of arguments, so you should only do it when:</p>
<ul>
<li>Performance of this piece of code absolutely matters.</li>
<li>You don&rsquo;t know the types you&rsquo;re dealing with (generic code) or they are expensive to move (i.e. because they don&rsquo;t have a move constructor).</li>
<li>You have only few parameters.</li>
</ul>
<p>With the <code>person</code> example, performance is probably not critical and we know that <code>std::string</code> is cheap to move, so there is no need to duplicate the constructor.</p>
<p>But with <code>std::vector&lt;T&gt;::push_back()</code> all three conditions are true, which is why the two overloads exist.</p>
<h3>2. Rvalue Reference Parameter For Conditional Moving</h3>
<p>Move is not just an optimization of copy, it has an important semantic meaning &mdash; transferring ownership.</p>
<p>Consider a function that wants to take ownership from a <code>std::unique_ptr</code>. In order to do that it needs to move from the argument.</p>
<p>This gives three options for the parameter type:</p>
<ul>
<li><code>std::unique_ptr&lt;T&gt;&amp;</code></li>
<li><code>std::unique_ptr&lt;T&gt;</code></li>
<li><code>std::unique_ptr&lt;T&gt;&amp;&amp;</code></li>
</ul>
<p>Taking an lvalue reference is a bad idea: it is not obvious for the caller that the pointer is going to be in the moved-from state after the function returns. Also the function cannot take real rvalues &mdash; temporaries.</p>
<p>Taking it by value works as now the caller has to write <code>std::move()</code> when moving an existing an object (lvalue). This comes with an extra move operation again, but this is usually negligible.</p>
<p>An rvalue reference parameter is the same for the caller, but without the extra move internally, so it seems superior. But consider this function:</p>
<div>
<pre tabindex="0"><code><span><span><span>void</span> <span>foo</span><span>(</span><span>std</span><span>::</span><span>unique_ptr</span><span>&lt;</span><span>T</span><span>&gt;&amp;&amp;</span><span>)</span> <span>{</span> <span>/* do nothing */</span> <span>}</span>
</span></span></code></pre>
</div>
<p>This function will not actually move from the argument, so it will stay the same after the call!</p>
<p>An rvalue reference parameter here thus has slightly different semantics: it allows for <em>conditional</em> moving. Maybe the function will take ownership, maybe not.</p>
<p>However, this use case is pretty rare.</p>
<h3>3. Rvalue Reference Parameters For Forcing Move</h3>
<p>An rvalue reference parameter has an additional difference to a by value parameter: It forces the caller to use <code>std::move()</code> for lvalues, even if the type is not really moveable!</p>
<div>
<pre tabindex="0"><code><span><span><span>void</span> <span>foo</span><span>(</span><span>int</span><span>&amp;&amp;</span> <span>i</span><span>);</span>
</span></span><span><span><span>&hellip;</span>
</span></span><span><span><span>auto</span> <span>i</span> <span>=</span> <span>42</span><span>;</span>
</span></span><span><span><span>foo</span><span>(</span><span>i</span><span>);</span> <span>// error: i is an lvalue
</span></span></span><span><span><span>foo</span><span>(</span><span>std</span><span>::</span><span>move</span><span>(</span><span>i</span><span>));</span> <span>// okay
</span></span></span></code></pre>
</div>
<p>But this can be used for our advantage: In a new project I&rsquo;m working I have the following (simplified) piece of code.</p>
<div>
<pre tabindex="0"><code><span><span><span>// some block of memory
</span></span></span><span><span><span>struct</span> <span>memory_block</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>std</span><span>::</span><span>byte</span><span>*</span> <span>address</span><span>;</span>
</span></span><span><span> <span>std</span><span>::</span><span>size_t</span> <span>size</span><span>;</span>
</span></span><span><span><span>};</span>
</span></span><span><span>
</span></span><span><span><span>memory_block</span> <span>new_block</span><span>(</span><span>std</span><span>::</span><span>size_t</span> <span>size</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>// allocate block using ::operator new
</span></span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>void</span> <span>delete_block</span><span>(</span><span>memory_block</span><span>&amp;&amp;</span> <span>block</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>// deallocate block using ::operator delete
</span></span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>Note that <code>delete_block()</code> takes the <code>memory_block</code> by rvalue reference &mdash; even though it is a trivially copyable type. But forcing the caller to write <code>delete_block(std::move(block))</code> makes it really obvious that the memory block is being turned unusable.</p>
<p>I&rsquo;ve only started using this recently, so I don&rsquo;t have much experience with it. But I encourage you to experiment.</p>
<h2>Rvalue Ref-qualified Member Functions</h2>
<p>C++11 added a relatively obscure feature, ref-qualifiers for <code>*this</code>.</p>
<blockquote>
<p>With &ldquo;obscure&rdquo; I mean: if you&rsquo;re not writing libraries, you&rsquo;ve probably haven&rsquo;t heard of it.</p>
</blockquote>
<p>You could always qualify member functions with <code>const</code>, to allow them being called on <code>const</code> objects.</p>
<blockquote>
<p>You could also always qualify them with <code>volatile</code>, but &hellip; <code>volatile</code>.</p>
</blockquote>
<p>Similar, in C++11 you can qualify them with <code>&amp;</code> and <code>&amp;&amp;</code> as well. Due to backwards compatibility, they have slightly different rules:</p>
<ul>
<li>A member function qualified with <code>&amp;</code> cannot be called on rvalues (unless it is also qualified with <code>const</code>).</li>
<li>A member function qualified with <code>&amp;&amp;</code> cannot be called on lvalues.</li>
<li>A member function not qualified can be called on lvalues and rvalues.</li>
</ul>
<p>Usually, you don&rsquo;t have just a single member function qualified with <code>&amp;&amp;</code>, for example, but have multiple overloads with the different qualification. The combination of overloads depends on the use case.</p>
<h3>1. Passing The Value Category Along In A Getter Function</h3>
<p>Consider <code>std::optional&lt;T&gt;</code>: It (maybe) contains an object of type <code>T</code> and provides access to it with <code>.value()</code>.</p>
<blockquote>
<p>It also has <code>operator*()</code>, but I always pretend it doesn&rsquo;t exist.</p>
</blockquote>
<p>Like usually, it gives you a <code>const T</code> when you call it on a <code>const</code> object, and a non-<code>const</code> <code>T</code> when you call it on a non-<code>const</code> object:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>optional</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>opt</span><span>(</span><span>42</span><span>);</span>
</span></span><span><span><span>opt</span><span>.</span><span>value</span><span>()</span> <span>=</span> <span>43</span><span>;</span> <span>// okay
</span></span></span><span><span>
</span></span><span><span><span>const</span> <span>std</span><span>::</span><span>optional</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>opt</span><span>(</span><span>42</span><span>);</span>
</span></span><span><span><span>opt</span><span>.</span><span>value</span><span>()</span> <span>=</span> <span>43</span><span>;</span> <span>// error
</span></span></span></code></pre>
</div>
<p>And - you&rsquo;ve guessed it - it gives you a <code>T&amp;&amp;</code> when called on an rvalue and a <code>T&amp;</code> when called on an lvalue.</p>
<p>Why was that done?</p>
<p>Consider this:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>optional</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>make</span><span>();</span>
</span></span><span><span>
</span></span><span><span><span>&hellip;</span>
</span></span><span><span>
</span></span><span><span><span>std</span><span>::</span><span>string</span> <span>str</span> <span>=</span> <span>make</span><span>().</span><span>value</span><span>();</span>
</span></span></code></pre>
</div>
<p>Here <code>make()</code> returns an rvalue optional, so it is safe to move the string out. And this is exactly what is done, thanks to the rvalue qualified <code>value()</code> that is being called!</p>
<p>In order to achieve this use case, there are four overloads of <code>std::optional&lt;T&gt;::value()</code>, one for each combination of <code>const</code> and rvalue-ness:</p>
<div>
<pre tabindex="0"><code><span><span><span>// assume a `get_pointer()` function that returns a pointer to the object being stored
</span></span></span><span><span>
</span></span><span><span><span>T</span><span>&amp;</span> <span>value</span><span>()</span> <span>&amp;</span> <span>// non-const lvalue
</span></span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>*</span><span>get_pointer</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>T</span><span>&amp;&amp;</span> <span>value</span><span>()</span> <span>&amp;&amp;</span> <span>// non-const rvalue
</span></span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>*</span><span>get_pointer</span><span>());</span> <span>// propagate rvalue-ness
</span></span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>const</span> <span>T</span><span>&amp;</span> <span>value</span><span>()</span> <span>const</span> <span>&amp;</span> <span>// const lvalue
</span></span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>*</span><span>get_pointer</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span><span><span>
</span></span><span><span><span>const</span> <span>T</span><span>&amp;&amp;</span> <span>value</span><span>()</span> <span>const</span> <span>&amp;&amp;</span> <span>// const rvalue
</span></span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>*</span><span>get_pointer</span><span>());</span> <span>// propagate rvalue-ness
</span></span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<blockquote>
<p>If you think &ldquo;the last combination is somewhat useless&rdquo; - you&rsquo;re right. But not having it <a href="https://cplusplus.github.io/LWG/issue2485">breaks weird code</a>.</p>
</blockquote>
<p>While this use-case is cool, unless you&rsquo;re writing types like <code>std::optional</code>, you probably don&rsquo;t need it.</p>
<h3>2. Optimize Member Functions</h3>
<p>But maybe a more general idea is useful: You have a <code>const</code> member function, that does some expensive calculation, maybe it needs to copy internal state for the result:</p>
<div>
<pre tabindex="0"><code><span><span><span>expensive_state</span> <span>foo</span><span>(</span><span>int</span> <span>arg</span><span>)</span> <span>const</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>expensive_state</span> <span>copy</span> <span>=</span> <span>my_state_</span><span>;</span>
</span></span><span><span> <span>do_sth</span><span>(</span><span>copy</span><span>,</span> <span>arg</span><span>);</span> </span></span><span><span> <span>return</span> <span>copy</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>An rvalue qualified overload can get rid of the copy by using the internal state directly - after all the object is a temporary and/or no longer needed, so no need to keep it intact:</p>
<div>
<pre tabindex="0"><code><span><span><span>expensive_state</span><span>&amp;&amp;</span> <span>foo</span><span>(</span><span>int</span> <span>arg</span><span>)</span> <span>&amp;&amp;</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>do_sth</span><span>(</span><span>my_state_</span><span>,</span> <span>arg</span><span>);</span>
</span></span><span><span> <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>my_state_</span><span>);</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>When optimizing certain functions for rvalues, you usually have two overloads: one <code>const</code> and one <code>&amp;&amp;</code> qualified. Make sure the optimization is worth the code duplication though!</p>
<h3>3. Prevent Dangling References</h3>
<p>I&rsquo;ve talked about it in detail in <a href="https://www.foonathan.net/blog/2017/03/22/string_view-temporary.html">this blog post</a>: if <code>std::string_view</code> is used outside of a function parameter, be very careful!</p>
<p>For example, consider this simple getter:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string_view</span> <span>get_name</span><span>()</span> <span>const</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>name_</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>Over time it is decided that the name should be split into first and last name.</p>
<blockquote>
<p>Even though <a href="https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/">this is not a good idea</a>.</p>
</blockquote>
<p>In a late night refactor the getter is changed:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string_view</span> <span>get_name</span><span>()</span> <span>const</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>first_name_</span> <span>+</span> <span>last_name_</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>This blows up, because now we&rsquo;ve created a view to a temporary string!</p>
<p>One way this could have been prevented is by disabling the conversion to <code>std::string_view</code> for rvalues. Right now we have this:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string</span><span>::</span><span>operator</span> <span>std</span><span>::</span><span>string_view</span><span>()</span> <span>const</span> <span>{</span> <span>return</span> <span>&hellip;</span><span>;</span> <span>}</span>
</span></span></code></pre>
</div>
<blockquote>
<p>Ignoring the templates here.</p>
</blockquote>
<p>By adding a second overload, we can prevent the use for temporaries:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string</span><span>::</span><span>operator</span> <span>std</span><span>::</span><span>string_view</span><span>()</span> <span>const</span> <span>&amp;</span> <span>{</span> <span>return</span> <span>&hellip;</span><span>;</span> <span>}</span>
</span></span><span><span><span>std</span><span>::</span><span>string</span><span>::</span><span>operator</span> <span>std</span><span>::</span><span>string_view</span><span>()</span> <span>const</span> <span>&amp;&amp;</span> <span>=</span> <span>delete</span><span>;</span>
</span></span></code></pre>
</div>
<p>That way, overload resolution will select this overload when called on an rvalue, then issue an error because the function is deleted.</p>
<p>Read more about deleting functions in <a href="https://www.foonathan.net/blog/2015/10/16/overload-resolution-1.html">my series about controlling overload resolution</a>.</p>
<h3>4. Mark One Time Operations that Destroy the Objects</h3>
<p>I recently have a lot of immutable objects, so I use the builder pattern a lot:</p>
<div>
<pre tabindex="0"><code><span><span><span>class</span> <span>Foo</span>
</span></span><span><span><span>{</span>
</span></span><span><span><span>public</span><span>:</span>
</span></span><span><span> <span>class</span> <span>Builder</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>public</span><span>:</span>
</span></span><span><span> <span>Builder</span><span>()</span> <span>=</span> <span>default</span><span>;</span> </span></span><span><span>
</span></span><span><span> <span>void</span> <span>set_bar</span><span>(</span><span>Bar</span> <span>b</span><span>)</span> <span>{</span> <span>&hellip;</span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>void</span> <span>set_baz</span><span>(</span><span>Baz</span> <span>b</span><span>)</span> <span>{</span> <span>&hellip;</span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>Foo</span><span>&amp;&amp;</span> <span>finish</span><span>()</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>obj</span><span>);</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>private</span><span>:</span>
</span></span><span><span> <span>Foo</span> <span>obj_</span><span>;</span>
</span></span><span><span> <span>};</span>
</span></span><span><span>
</span></span><span><span><span>private</span><span>:</span>
</span></span><span><span> <span>Foo</span><span>()</span> <span>=</span> <span>default</span><span>;</span>
</span></span><span><span><span>};</span>
</span></span></code></pre>
</div>
<p>Note the <code>finish()</code> function: once the object is done it will be moved out. But this will destroy the builder, i.e. you can&rsquo;t use it again anymore.</p>
<p>Granted, with a method name <code>finish()</code> it may be obvious. But still, qualify the member function with <code>&amp;&amp;</code>:</p>
<div>
<pre tabindex="0"><code><span><span><span>Foo</span><span>&amp;&amp;</span> <span>finish</span><span>()</span> <span>&amp;&amp;</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>return</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>obj</span><span>);</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>Then you will be reminded that the object is turned unusable:</p>
<div>
<pre tabindex="0"><code><span><span><span>auto</span> <span>obj</span> <span>=</span> <span>builder</span><span>.</span><span>finish</span><span>();</span> <span>// error!
</span></span></span><span><span><span>auto</span> <span>obj2</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>builder</span><span>).</span><span>finish</span><span>();</span> <span>// okay
</span></span></span></code></pre>
</div>
<h2>Rvalue References As Return Types</h2>
<p>When using rvalue references as return types, returning temporaries or function local variables still creates a dangling reference, just like with lvalue references.</p>
<p>So it is only really applicable if you&rsquo;re returning something like member variables or reference parameters and you&rsquo;re moving them out of place.</p>
<p>There is only one function that moves a reference parameter out: <code>std::move</code>.</p>
<p>But we&rsquo;ve already seen a couple of function that move member variables out: The recent <code>Builder::finish()</code> as well as the <code>optional::value()</code> for rvalues. Both of those could return it by value, so should they use an rvalue reference?</p>
<p>It&rsquo;s complicated.</p>
<p>Returning an rvalue in the <code>optional::value()</code> case ensures that <code>decltype()</code> is consistent with the other cases. However, it can lead to dangling references if you do something like this:</p>
<div>
<pre tabindex="0"><code><span><span><span>optional</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>foo</span><span>();</span>
</span></span><span><span>
</span></span><span><span><span>auto</span><span>&amp;&amp;</span> <span>val</span> <span>=</span> <span>foo</span><span>().</span><span>value</span><span>();</span>
</span></span></code></pre>
</div>
<p>The temporary returned by the function is destroyed, so <code>val</code> references the destroyed member variable. However, if <code>value()</code> returned a <code>T</code> by value, lifetime extension would ensure that the reference is still valid.</p>
<p>On the other hand, returning by reference may save one extra move.</p>
<p>So should you use rvalue return types?</p>
<p>I&rsquo;d say you should only do it if you have an rvalue qualified member function but think about the consequences.</p>
<h2>Rvalue References As Class Members</h2>
<p>Don&rsquo;t put references into classes, it will make it hard to write an <code>operator=</code>.</p>
<p>Instead store a pointer &mdash; or better yet &mdash; something which is never null like my <code>type_safe::object_ref&lt;T&gt;</code>.</p>
<h2>Conclusion</h2>
<p>Use rvalue references as function parameters for conditional moving, to force callers to write <code>std::move()</code>, and &mdash; together with a <code>const T&amp;</code> overload &mdash; for more efficient input parameters.</p>
<p>Use ref-qualified member functions to mirror value categories in getters, optimize member functions, prevent operations on temporaries, and to mark member functions that destroy the object&rsquo;s internal state.</p>
<p>Be careful about rvalue reference return types and don&rsquo;t put them into classes.</p>
<blockquote>
<p>If you've liked this blog post, consider <a href="https://paypal.me/foonathan">donating</a> or otherwise <a href="https://jonathanmueller.dev/support-me/" target="_blank">supporting me</a>.</p>
</blockquote>
<blockquote>
<p>This blog post was written for my old blog design and ported over. If there are any issues, please <a href="https://jonathanmueller.dev">let me know</a>.</p>
</blockquote>
<footer>
<div>
<section>
<figure><img src="https://gravatar.com/avatar/9cbeac7efd2dd7b3f2235df7e45d11fc?s=80"></figure>
<div>
<p>C++ enthusiast. I write libraries.</p>
</div>
</section>
</div>
</footer>
</article>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via foonathan::&#8203;blog()<br>
<br>
September 1, 2022 at 03:52PM
